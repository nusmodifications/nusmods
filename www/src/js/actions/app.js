// @flow

import type { FSA } from 'types/redux';
import type { NotificationOptions } from 'types/reducers';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}

export const TOGGLE_FEEDBACK_MODAL = 'TOGGLE_FEEDBACK_MODAL';
export function toggleFeedback(): FSA {
  return {
    type: TOGGLE_FEEDBACK_MODAL,
    payload: null,
  };
}

export const OPEN_NOTIFICATION = 'OPEN_NOTIFICATION';
export function openNotification(message: string, options: NotificationOptions = {}): FSA {
  return {
    type: OPEN_NOTIFICATION,
    payload: {
      message,
      ...options,
    },
  };
}

export const POP_NOTIFICATION = 'POP_NOTIFICATION';
export function popNotification(): FSA {
  return {
    type: POP_NOTIFICATION,
    payload: null,
  };
}
