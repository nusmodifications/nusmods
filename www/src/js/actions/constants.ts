/** moduleBank constants * */

export const FETCH_MODULE = 'FETCH_MODULE'; // Action to fetch modules
export const FETCH_MODULE_LIST = 'FETCH_MODULE_LIST';
export const UPDATE_MODULE_TIMESTAMP = 'UPDATE_MODULE_TIMESTAMP';
export const REMOVE_LRU_MODULE = 'REMOVE_LRU_MODULE';
export const FETCH_ARCHIVE_MODULE = 'FETCH_ARCHIVE_MODULE'; // Action to fetch module from previous years

/** undoHistory constants * */

export const UNDO = 'UNDO';
export const REDO = 'REDO';

/** export constant(s) * */

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
