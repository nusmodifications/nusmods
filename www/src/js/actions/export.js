// @flow
import type { Module, Semester } from 'types/modules';

import { hydrateSemTimetableWithLessons } from 'utils/timetables';
import type { ExportData } from 'types/export';
import type { FSA, GetState } from 'types/redux';
import { getSemesterTimetable } from 'reducers/timetables';

function downloadUrl(blob: Blob, filename: string) {
  const link = document.createElement('a');
  if (window.navigator.msSaveOrOpenBlob) {
    window.navigator.msSaveOrOpenBlob(blob, filename);
  } else {
    const objectUrl = URL.createObjectURL(blob);
    link.download = filename;
    link.href = objectUrl;
    link.dispatchEvent(new MouseEvent('click'));
    URL.revokeObjectURL(objectUrl);
  }
}

export const SUPPORTS_DOWNLOAD = 'download' in document.createElement('a');

export function downloadAsIcal(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    Promise.all([
      import(/* webpackChunkName: "export" */ 'ical-generator'),
      import(/* webpackChunkName: "export" */ 'utils/ical'),
    ]).then(([ical, icalUtils]) => {
      const { moduleBank, timetables } = getState();
      const modules = moduleBank.modules;
      const { timetable } = getSemesterTimetable(semester, timetables);
      const timetableWithLessons = hydrateSemTimetableWithLessons(timetable, modules, semester);

      const events = icalUtils.default(semester, timetableWithLessons, modules);
      // $FlowFixMe Flow doesn't seem to like import() for CJS modules
      const cal = ical({
        domain: 'nusmods.com',
        prodId: '//NUSMods//NUSMods//EN',
        events,
      });

      const blob = new Blob([cal.toString()], { type: 'text/plain' });
      downloadUrl(blob, 'nusmods_calendar.ics');
    });
  };
}

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
export function setExportedData(modules: Module[], data: ExportData): FSA {
  return {
    type: SET_EXPORTED_DATA,
    payload: {
      modules,
      ...data,
    },
  };
}
