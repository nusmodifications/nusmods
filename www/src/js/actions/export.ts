// @flow

import type { Module, Semester } from 'types/modules';
import type { ExportData } from 'types/export';
import type { FSA, GetState } from 'types/redux';
import { hydrateSemTimetableWithLessons } from 'utils/timetables';
import { captureException, retryImport } from 'utils/error';
import { getSemesterTimetable } from 'reducers/timetables';

function downloadUrl(blob: Blob, filename: string) {
  const link = document.createElement('a');
  if (window.navigator.msSaveOrOpenBlob) {
    window.navigator.msSaveOrOpenBlob(blob, filename);
  } else {
    const objectUrl = URL.createObjectURL(blob);
    link.download = filename;
    link.href = objectUrl;
    link.dispatchEvent(new MouseEvent('click'));
    URL.revokeObjectURL(objectUrl);
  }
}

export const SUPPORTS_DOWNLOAD = 'download' in document.createElement('a');

export function downloadAsIcal(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    Promise.all([
      retryImport(() => import(/* webpackChunkName: "export" */ 'ical-generator')),
      retryImport(() => import(/* webpackChunkName: "export" */ 'utils/ical')),
    ])
      .then(([ical, icalUtils]) => {
        const {
          moduleBank: { modules },
          timetables,
        } = getState();
        const { timetable } = getSemesterTimetable(semester, timetables);
        const timetableWithLessons = hydrateSemTimetableWithLessons(timetable, modules, semester);

        const events = icalUtils.default(semester, timetableWithLessons, modules);
        const cal = ical.default({
          domain: 'nusmods.com',
          prodId: '//NUSMods//NUSMods//EN',
          events,
        });

        const blob = new Blob([cal.toString()], { type: 'text/plain' });
        downloadUrl(blob, 'nusmods_calendar.ics');
      })
      .catch(captureException);
  };
}

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
export function setExportedData(modules: Module[], data: ExportData): FSA {
  return {
    type: SET_EXPORTED_DATA,
    payload: {
      modules,
      ...data,
    },
  };
}
