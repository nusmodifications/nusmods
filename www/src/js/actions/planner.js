// @flow

import type { ModuleCode, Semester } from 'types/modules';
import type { FSA } from 'types/redux';

export const ADD_PLANNER_YEAR = 'ADD_PLANNER_YEAR';
export function addPlannerYear(year: string): FSA {
  return {
    type: ADD_PLANNER_YEAR,
    payload: year,
  };
}

export const ADD_PLANNER_MODULE = 'ADD_PLANNER_MODULE';
export function addPlannerModule(
  moduleCode: ModuleCode,
  year: string,
  semester: Semester,
  index: ?number = null,
): FSA {
  return {
    type: ADD_PLANNER_MODULE,
    payload: {
      year,
      semester,
      moduleCode,
      index,
    },
  };
}

export const MOVE_PLANNER_MODULE = 'MOVE_PLANNER_MODULE';
export function movePlannerModule(
  moduleCode: ModuleCode,
  year: string,
  semester: Semester,
  index: ?number = null,
): FSA {
  return {
    type: MOVE_PLANNER_MODULE,
    payload: {
      year,
      semester,
      moduleCode,
      index,
    },
  };
}

export const REMOVE_PLANNER_MODULE = 'REMOVE_PLANNER_MODULE';
export function removePlannerModule(moduleCode: ModuleCode): FSA {
  return {
    type: REMOVE_PLANNER_MODULE,
    payload: {
      moduleCode,
    },
  };
}
