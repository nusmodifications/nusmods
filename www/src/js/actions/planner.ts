// @flow

import type { ModuleCode, Semester } from 'types/modules';
import type { FSA } from 'types/redux';
import type { CustomModule } from 'types/reducers';

export const SET_PLANNER_MIN_YEAR = 'SET_PLANNER_MIN_YEAR';
export function setPlannerMinYear(year: string) {
  return {
    type: SET_PLANNER_MIN_YEAR,
    payload: year,
  };
}

export const SET_PLANNER_MAX_YEAR = 'SET_PLANNER_MAX_YEAR';
export function setPlannerMaxYear(year: string) {
  return {
    type: SET_PLANNER_MAX_YEAR,
    payload: year,
  };
}

export const SET_PLANNER_IBLOCS = 'SET_PLANNER_IBLOCS';
export function setPlannerIBLOCs(iblocs: boolean) {
  return {
    type: SET_PLANNER_IBLOCS,
    payload: iblocs,
  };
}

export const ADD_PLANNER_MODULE = 'ADD_PLANNER_MODULE';
export function addPlannerModule(
  moduleCode: ModuleCode,
  year: string,
  semester: Semester,
  index: ?number = null,
): FSA {
  return {
    type: ADD_PLANNER_MODULE,
    payload: {
      year,
      semester,
      moduleCode,
      index,
    },
  };
}

export const MOVE_PLANNER_MODULE = 'MOVE_PLANNER_MODULE';
export function movePlannerModule(
  moduleCode: ModuleCode,
  year: string,
  semester: Semester,
  index: ?number = null,
): FSA {
  return {
    type: MOVE_PLANNER_MODULE,
    payload: {
      year,
      semester,
      moduleCode,
      index,
    },
  };
}

export const REMOVE_PLANNER_MODULE = 'REMOVE_PLANNER_MODULE';
export function removePlannerModule(moduleCode: ModuleCode): FSA {
  return {
    type: REMOVE_PLANNER_MODULE,
    payload: {
      moduleCode,
    },
  };
}

export const ADD_CUSTOM_PLANNER_DATA = 'ADD_CUSTOM_PLANNER_DATA';
export function addCustomModule(moduleCode: ModuleCode, data: CustomModule): FSA {
  return {
    type: ADD_CUSTOM_PLANNER_DATA,
    payload: { moduleCode, data },
  };
}
