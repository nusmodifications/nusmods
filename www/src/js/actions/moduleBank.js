// @flow
import type { AcadYear, Module, ModuleCode } from 'types/modules';
import type { TimetableConfig } from 'types/timetables';

import type { GetState } from 'types/redux';
import type { ModulesMap } from 'reducers/moduleBank';
import { size, get, flatMap, sortBy, zip } from 'lodash';
import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';

const MAX_MODULE_LIMIT: number = 100;

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

// Action to fetch modules
export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ?ModuleCode {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export const UPDATE_MODULE_TIMESTAMP = 'UPDATE_MODULE_TIMESTAMP';
export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export const REMOVE_LRU_MODULE = 'REMOVE_LRU_MODULE';
export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result) => {
        onFinally();
        return result;
      },
      (error) => {
        onFinally();
        throw error;
      },
    );
  };
}

// Action to fetch module from previous years
export const FETCH_ARCHIVE_MODULE: string = 'FETCH_ARCHIVE_MODULE';
export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Array<Module>>(config.archiveYears, modules));
}
