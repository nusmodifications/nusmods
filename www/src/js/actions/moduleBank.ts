import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';
import { flatMap, get, size, sortBy, zip } from 'lodash';
import { ModulesMap } from 'types/reducers';
import { AcadYear, Module, ModuleCode } from 'types/modules';

import { GetState } from 'types/redux';
import { TimetableConfig } from 'types/timetables';
import {
  FETCH_ARCHIVE_MODULE,
  FETCH_MODULE,
  FETCH_MODULE_LIST,
  REMOVE_LRU_MODULE,
  UPDATE_MODULE_TIMESTAMP,
} from './constants';

const MAX_MODULE_LIMIT = 100;

export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ModuleCode | null {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove: ModuleCode[] = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy<ModuleCode>(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result: any) => {
        onFinally();
        return result;
      },
      (error: any) => {
        onFinally();
        throw error;
      },
    );
  };
}

export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Module[]>(config.archiveYears, modules));
}
