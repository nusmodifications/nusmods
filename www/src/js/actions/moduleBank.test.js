// @flow
import type { FSA } from 'types/redux';

import * as actions from 'actions/moduleBank';

test('fetchModuleList should return a request action', () => {
  const resultOfAction: FSA = actions.fetchModuleList();
  expect(resultOfAction).toMatchSnapshot();
});

test('fetchModule should return a request action', () => {
  const resultOfAction: FSA = actions.fetchModule('CS1010S');
  expect(resultOfAction).toMatchSnapshot();
});
