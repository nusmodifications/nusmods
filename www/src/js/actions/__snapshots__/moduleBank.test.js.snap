// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchModule should return a request action 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "test://MOCK_MOD_DETAILS_URL/CS1010S",
  },
  "type": "FETCH_MODULE",
}
`;

exports[`fetchModuleList should return a request action 1`] = `
Object {
  "meta": Object {
    Symbol(API_REQUEST): true,
  },
  "payload": Object {
    "method": "GET",
    "url": "test://MOCK_MOD_LIST_URL",
  },
  "type": "FETCH_MODULE_LIST",
}
`;
